processing/calculations.py

```diff
@@ -0,0 +1,122 @@
+                component_metrics[key] = [value]
+            else:
+                component_metrics[key].append(value)
+
+    # prepare output dict
+    output_dict = {'success': process_dict["success"],
+                   'process': process_dict["process"],
+                   'actual_target_metrics': {}}
+
+    calculations = output_dict['actual_target_metrics']
+
+    for key, value in component_metrics.items():
+        calculations[key] = {}
+
+        # calculate actual metrics
+        calculations[key]['actual'] = {"total": sum(value),
+                                       "min": min(value),
+                                       "max": max(value),
+                                       "avg": mean(value),
+                                       }
+
+        # more than one component has this metric, thus a standard deviation can be calculated
+        if len(value) > 1:
+            calculations[key]['actual'].update({"std_dev": stdev(value)})
+
+        # only one component has this metric, thus no standard deviation can be calculated
+        else:
+            calculations[key]['actual'].update({"std_dev": None})
+
+        # get amount of components
+        calculations[key]["count_component"] = len(value)
+
+        # calculate and get target metrics
+        calculations[key]['target'] = {'avg': process_dict['target_metrics'][key],
+                                       'total': process_dict['target_metrics'][key] * len(value)}
+
+    return output_dict
+
+
+def compare_actual_target_metrics(process_dict: dict, metrics_dict: dict) -> dict:
+    """
+    Compares the actual value against the target value
+    and sets the fulfillment to true or false based on
+    the comparator of the related metric.
+
+    :param process_dict: output from calculate_current_values()
+    :type process_dict: dict
+    :param metrics_dict: output from get_metrics_data()
+    :type metrics_dict: dict
+    :return: process_dict
+    """
+
+    for metric in process_dict['actual_target_metrics']:
+        comparator = metrics_dict['metrics'][metric]['fulfilled_if']
+        process_metrics_dict = process_dict['actual_target_metrics'][metric]
+
+        if eval(f"{process_metrics_dict['actual']['avg']} {comparator}"
+                f"{process_metrics_dict['target']['avg']}"):
+
+            fulfillment = True
+        else:
+            fulfillment = False
+
+        process_dict['actual_target_metrics'][metric]['fulfillment'] = fulfillment
+
+    return process_dict
+
+
+def calculate_risk_score(process_dict: dict) -> dict:
+    """
+    Calculates the average fulfillment rate for
+    all compared metrics
+
+    :param process_dict: from compare_actual_target_metrics()
+    :type process_dict: dict
+    :return: process_dict
+    """
+
+    sum = 0
+    sub_dict = process_dict["actual_target_metrics"]
+    for metric in sub_dict:
+        sum += sub_dict[metric]["fulfillment"]
+
+    process_dict["score"] = int((sum / len(sub_dict)) * 100)
```

Comment:
Bei mir schmeisst er bei einem Aufruf aus dem Core über get_process noch einen Fehler auf, weil die Länge des sub_dict wohl null ist und man nicht durch null teilen kann. Ich weiss nicht ob der Error auch in der Produktion mit "richtigen" Prozessen auftreten kann oder es durch einen Fehler der Vorgängerfunktion dazu kommt... Vlt. guckt ihr da nochmal kurz drauf.

Reply:
Habe drei Dinge geändert:
1) Check, ob "actual_target_metrics" im Dict existiert,
2) Check, ob "fulfillment" im jeweiligen Sub-Dict existiert,
3) Ein max(len(dict), 1), sodass immer mindestens durch 1 geteilt wird.

Reply:
Wann wäre das denn überhaupt der Fall in einem korrekten Beispiel? Ansonsten: es soll ja extra einen Fehler werfen und nicht weiterlaufen (bspw. wenn "actual_target_metrics" nicht existiert oder darin keine Metriken sind oder diese kein fulfillment enthalten), dafür haben wir ja den error_handler in endpoints erstellt, da sonst das Frontend in einen Fehler bei der Verarbeitung läuft. Vom Frontend werden entweder korrekte Werte oder ein Fehler 500 erwartet.

Reply:
Danke fürs Beheben. Für mich sieht das gut aus und wir sollten zumindest an der Stellen nicht mehr in einen Fehler laufen können.👍🏼

@Timo-Buecher , @NimsaJ150 Vlt. sollten wir das mal generell definieren wie wir sowas organisatorisch lösen. Schöner wäre es natürlich wenn das von den Vorgängerfunktion gelöst werden könnte.

Reply:
Meiner Meinung nach müssen wir an der Stelle zumindest für 1. und 2. in einen Fehler laufen. Siehe Anmerkung oben. Ich habe bereits in einem JF vorgestellt, wie wir mit Fehlern umgehen. Gerne können wir darüber nochmal reden, falls das nicht passen sollte, oder ich es auch schriftlich ausformulieren.

Über den Fall, dass keine actual metrics gegeben sind schaue ich nochmal drüber.